### ISSUE TITLE
White-list build images

### ISSUE THREAD

**Issue by Cadair** [drone/drone#1061]
I have recently started testing drone and it is great! I want to know how to restrict the containers that drone can execute, because as it appears to be set up at the moment anyone can modify the `.drone.yml` file and download and execute any container from the main docker registry, which I don't really want to enable on my build bot.

Is it possible to limit the containers that can be executed to either a whitelist, or just to disable download from the registry, so the containers have to be locally built / cached?

Thanks

**Comment by bradrydzewski** [comment]
It is not possible today, but I agree we should add this capability in our next release

**Comment by Cadair** [comment]
@bradrydzewski yes please! Do you know if there is a work around, like disabling docker from downloading from the registry?

**Comment by bradrydzewski** [comment]
@Cadair there is nothing you can configure inside of Drone, however, you can probably block `index.docker.io` at the linux level. For example, you could modify `/etc/hosts` or `/etc/hosts.deny` to block or redirect the index.

**Comment by Cadair** [comment]
@bradrydzewski thanks a lot.

I would offer a patch, but I have never written a line of Go in my life, and now is not the time for me to start!

**Comment by Cadair** [comment]
Thinking about this a lot more, a lot of the security issues I can think of, such as being able to run code as root in a container, and therefore potentially exploit docker to break out of the container, come from the ability to modify the `.drone.yml` file and then send a PR to the repo where the drone server will execute whatever code is called from the `.drone.yml` file.

I wonder therefore if it would be worth having an option to disable or at least suspend builds for PRs that modify the build script.

**Comment by bradrydzewski** [comment]
@Cadair there is a switch in the repository settings that let's you disable pull requests.

Drone mitigates some issues related to malicious pull requests, for example, by not injecting private variables into pull requests, disabling certain docker features in pull requests (privileged, net, etc) and more. But as you mentioned, a vulnerability in Docker could be exploited via a malicious pull request, in which case you may want to disable them for your public repos.

**Comment by Cadair** [comment]
@bradrydzewski yes, disabling the PR builds is something I am considering. Unfortunately they are really useful!!

Would it be possible to

**Comment by Cadair** [comment]
I can see a few more compromise options,

1. Allow PR builds from people in a GH org or only registered users.
1. Don't build a PR if the .drone.yml file has been changed.
1. Same as above but allow from a whitelist / org.

**Comment by bradrydzewski** [comment]
@Cadair some of this is being discussed in #849.  I'm open to these options.  I'm also interested to see how user namespaces impact the attack surface once it lands in Docker. Combined with Apparmor I expect this will get us pretty far, or at least to a point that I'm personally comfortable with, but I'll need to discuss more with the Docker folks

**Comment by donny-dont** [comment]
@bradrydzewski can we maybe get this in 0.5? Trying to appease our security team so we have a set of images that are OK.

**Comment by bradrydzewski** [comment]
We should check and see if there is a native solution to this in Docker. I think you can configure your Docker daemon to use a mirror, which may give you the ability through the mirror to limit which external images are pulled.

**Comment by bradrydzewski** [comment]
I'm going to close this particular issue as defined (White-list build images) because 0.5 mitigates many security concerns from a drone perspective and integration with user namespaces already has a separate issue.

For the ability to white-list build images I believe this should be handle at the docker daemon level. This seems like the perfect type of thing Docker will want to include in its enterprise container management / security / audit offerings. One also has the option to use dockerhub with mirroring to enforce more fine-grained control.

TLDR is that I think limiting installed containers is important but should be the responsibility of the docker daemon or some other tooling in the docker ecosystem since this transcends just drone.
